{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Serialization\n",
    "using Statistics\n",
    "using StatsBase\n",
    "using StatsPlots\n",
    "using CircularArrays\n",
    "using CSV, DataFrames\n",
    "using KernelDensity\n",
    "using Distributions\n",
    "using Random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Atgff = CSV.File(\"AP000423.gff\"; comment = \"#\", header = [\"accession\", \"software\", \"feature\", \"start\", \"stop\", \"score\", \"strand\", \"phase\", \"attributes\"]) |> DataFrame\n",
    "genome_length = first(Atgff[Atgff.feature .== \"region\", :stop])\n",
    "\n",
    "function rc(x::Real)\n",
    "    genome_length - x + 1\n",
    "end\n",
    "\n",
    "structuralRNA = sort!(filter(x->x.feature âˆˆ [\"tRNA\", \"rRNA\"], Atgff), :start)\n",
    "\n",
    "function filteroutstructuralRNA!(data, strand)\n",
    "    mask = filter(x->x.strand == strand, structuralRNA)\n",
    "    for rna in eachrow(mask)\n",
    "        if strand == \"+\"\n",
    "            data[rna.start:rna.stop] .= 0\n",
    "        else\n",
    "            data[rc(rna.stop):rc(rna.start)] .= 0\n",
    "        end\n",
    "    end\n",
    "    data\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "footprints_gff = CSV.File(\"Known_short_AP000423.gff\"; comment = \"#\", header = [\"accession\", \"software\", \"feature\", \"start\", \"stop\", \"score\", \"strand\", \"phase\", \"attributes\", \"present\"]) |> DataFrame\n",
    "filter!(x->occursin(\"present\", x.present) , footprints_gff)\n",
    "select!(footprints_gff, Not(:present))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 50\n",
    "N = 10000\n",
    "\n",
    "function log_normalise(fwd_data, rev_data)\n",
    "    log_fwd_data = log.(fwd_data .+ 1)\n",
    "    log_rev_data = log.(rev_data .+ 1)\n",
    "    log_fwd_data .-= min(minimum(log_fwd_data), minimum(log_rev_data))\n",
    "    log_fwd_data ./= max(maximum(log_fwd_data), maximum(log_rev_data))\n",
    "    log_rev_data .-= min(minimum(log_fwd_data), minimum(log_rev_data))\n",
    "    log_rev_data ./= max(maximum(log_fwd_data), maximum(log_rev_data))\n",
    "    log_fwd_data, log_rev_data\n",
    "end\n",
    "\n",
    "function sample_from_kde(data::Vector{Int}, N::Int; boundaries = (1,50))\n",
    "    h = KernelDensity.default_bandwidth(data)\n",
    "    sampled = Vector{Int}(undef, N)\n",
    "    i = 1\n",
    "    while i <= N\n",
    "        sample = round(Int, rand(Normal(rand(data), h)))\n",
    "        if first(boundaries) <= sample <= last(boundaries)\n",
    "            sampled[i] = sample\n",
    "            i += 1\n",
    "        end\n",
    "    end\n",
    "    sampled\n",
    "end\n",
    "\n",
    "function sample_from_kde(data::Vector{Float64}, N::Int; boundaries = (0.0,1.0))\n",
    "    h = KernelDensity.default_bandwidth(data)\n",
    "    sampled = Vector{Float64}(undef, N)\n",
    "    i = 1\n",
    "    while i <= N\n",
    "        sample = rand(Normal(rand(data), h))\n",
    "        if first(boundaries) <= sample <= last(boundaries)\n",
    "            sampled[i] = sample\n",
    "            i += 1\n",
    "        end\n",
    "    end\n",
    "    sampled\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "footprint_lengths = footprints_gff.stop .- footprints_gff.start .+ 1\n",
    "kde_lengths = kde(footprint_lengths, boundary=(1,window))\n",
    "plot(kde_lengths; color = :dodgerblue, label = \"observed footprint_lengths\")\n",
    "simulated_lengths = sample_from_kde(footprint_lengths, N; boundaries = (15,40))\n",
    "plot!(kde(simulated_lengths); color = :skyblue, label = \"simulated footprint_lengths\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fwd_conservation = CircularVector(deserialize(\"arabidopsis_conservation.bin\"))\n",
    "rev_conservation = reverse(fwd_conservation)\n",
    "fwd_conservation, rev_conservation = log_normalise(fwd_conservation, rev_conservation)\n",
    "\n",
    "conservation_footprints = Vector{Vector{Float64}}()\n",
    "for f in eachrow(footprints_gff)\n",
    "    midpoint = f.start + (f.stop - f.start)/2\n",
    "    if mod(f.stop - f.start, 2) == 0; midpoint += 0.5; end\n",
    "    frange = Int(midpoint - 24.5):Int(midpoint + 24.5)\n",
    "    push!(conservation_footprints, fwd_conservation[frange])\n",
    "end\n",
    "\n",
    "conservation_peak_std = mean(std.(getindex.(conservation_footprints, Ref(13:37))))\n",
    "conservation_flanks_std = (mean(std.(getindex.(conservation_footprints, Ref(1:12)))) + mean(std.(getindex.(conservation_footprints, Ref(38:50)))))/2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = mean.(getindex.(conservation_footprints, Ref(20:30)))\n",
    "simulated_conservation_peaks = sample_from_kde(peaks, N)\n",
    "plot(kde(peaks); label = \"conservation peaks\", color = :green, xguide = \"transformed conservation\", yguide = \"density\")\n",
    "plot!(kde(simulated_conservation_peaks); label = \"simulated conservation peaks\", color = :palegreen, xlims = (0,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flanks = (mean.(getindex.(conservation_footprints, Ref(1:5))) .+ mean.(getindex.(conservation_footprints, Ref(46:50)))) ./ 2.0 ./ peaks\n",
    "simulated_conservation_flanks = sample_from_kde(flanks, N)\n",
    "plot(kde(flanks); color = :crimson, label = \"observed conservation flanks\", xguide = \"proportion of peak height\", yguide = \"density\")\n",
    "plot!(kde(simulated_conservation_flanks); color = :tomato, label = \"simulated conservation flanks\", xlims = (0,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fwd_srna = deserialize(\"srna_fwd.bin\")\n",
    "rev_srna = deserialize(\"srna_rev.bin\")\n",
    "\n",
    "filteroutstructuralRNA!(fwd_srna, \"+\")\n",
    "filteroutstructuralRNA!(rev_srna, \"-\")\n",
    "\n",
    "fwd_srna, rev_srna = log_normalise(fwd_srna, rev_srna)\n",
    "\n",
    "srna_footprints = Vector{Vector{Float64}}()\n",
    "for f in eachrow(footprints_gff)\n",
    "    midpoint = f.start + (f.stop - f.start)/2\n",
    "    if mod(f.stop - f.start, 2) == 0; midpoint += 0.5; end\n",
    "    if f.strand == \"-\"; midpoint = rc(midpoint); end\n",
    "    frange = Int(midpoint - 24.5):Int(midpoint + 24.5)\n",
    "    push!(srna_footprints, f.strand == \"+\" ? fwd_srna[frange] : rev_srna[frange])\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot()\n",
    "for f in srna_footprints\n",
    "    plot!(f; color = :gray, alpha = 0.5, label = false, xguide = \"nucleotide\", yguide = \"transformed read depth\")\n",
    "end\n",
    "plot!()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "srna_peak_std = median(std.(getindex.(srna_footprints, Ref(13:37))))\n",
    "srna_flanks_std = (median(std.(getindex.(srna_footprints, Ref(1:12)))) + median(std.(getindex.(srna_footprints, Ref(38:50)))))/2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = mean.(getindex.(srna_footprints, Ref(20:30)))\n",
    "simulated_peaks_srna = sample_from_kde(peaks, N)\n",
    "plot(kde(peaks); label = \"observed sRNA peaks\", color = :green, xguide = \"transformed read depth\", yguide = \"density\")\n",
    "plot!(kde(simulated_peaks_srna); color = :palegreen, label = \"simulated sRNA peaks\", xlims = (0,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flanks = (mean.(getindex.(srna_footprints, Ref(1:5))) .+ mean.(getindex.(srna_footprints, Ref(46:50)))) ./ 2.0 ./ peaks\n",
    "simulated_descents_srna = sample_from_kde(flanks, N)\n",
    "plot(kde(flanks); color = :crimson, label = \"observed sRNA flank descents\", xguide = \"proportion of peak height\", yguide = \"density\")\n",
    "plot!(kde(simulated_descents_srna); color = :tomato, label = \"simulated sRNA flank descents\", xlims = (0,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fwd_fiveprime = CircularVector(deserialize(\"ligated_vs_untreated.fwd.five.bin\"))\n",
    "filteroutstructuralRNA!(fwd_fiveprime, \"+\")\n",
    "rev_fiveprime = CircularVector(deserialize(\"ligated_vs_untreated.rev.five.bin\"))\n",
    "filteroutstructuralRNA!(rev_fiveprime, \"-\")\n",
    "\n",
    "fwd_fiveprime, rev_fiveprime = log_normalise(fwd_fiveprime, rev_fiveprime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fiveprime_termini = filter(x->x>0, vcat(fwd_fiveprime, rev_fiveprime))\n",
    "plot(kde(fiveprime_termini); label = \"observed 5' termini\", color = :green, xguide = \"transformed read depth\", yguide = \"density\", xlims = (0,1))\n",
    "simulated_fiveprime_termini = sample_from_kde(fiveprime_termini.data, N)\n",
    "plot!(kde(simulated_fiveprime_termini); color = :palegreen, label = \"simulated 5' termini\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fwd_threeprime = CircularVector(deserialize(\"TAP_vs_untreated.fwd.three.bin\"))\n",
    "filteroutstructuralRNA!(fwd_threeprime, \"+\")\n",
    "rev_threeprime = CircularVector(deserialize(\"TAP_vs_untreated.rev.three.bin\"))\n",
    "filteroutstructuralRNA!(rev_threeprime, \"-\")\n",
    "\n",
    "fwd_threeprime, rev_threeprime = log_normalise(fwd_threeprime, rev_threeprime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "threeprime_termini = filter(x->x>0, vcat(fwd_threeprime, rev_threeprime))\n",
    "plot(kde(threeprime_termini); label = \"observed 3' termini\", color = :crimson, xguide = \"transformed read depth\", yguide = \"density\", xlims = (0,1))\n",
    "simulated_threeprime_termini = sample_from_kde(threeprime_termini.data, N)\n",
    "plot!(kde(simulated_threeprime_termini); color = :tomato, label = \"simulated 3' termini\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function make_footprint_in_window(i::Int, fiveprime::Bool, threeprime::Bool)\n",
    "    footprintlength = simulated_lengths[i]\n",
    "    footprintstart = floor(Int, (window - footprintlength)/2)\n",
    "    footprint = UnitRange(footprintstart, footprintstart + footprintlength - 1)\n",
    "    padding = rand(2:5)\n",
    "\n",
    "    #conservation\n",
    "    footprint_conservation = simulated_conservation_peaks[i]\n",
    "    conservation = fill(footprint_conservation, window)\n",
    "    background = footprint_conservation * simulated_conservation_flanks[i]\n",
    "    conservation[1:footprint.start-1+padding] .= background\n",
    "    conservation[footprint.stop+1-padding:window] .= background\n",
    "\n",
    "    #sRNA\n",
    "    peak_height = simulated_peaks_srna[i]\n",
    "    srna = fill(peak_height, window)\n",
    "    background = peak_height * simulated_descents_srna[i]\n",
    "    srna[1:footprint.start-1] .= background\n",
    "    srna[footprint.stop+1:window] .= background\n",
    "\n",
    "    #termini\n",
    "    termini = zeros(Float64, window)\n",
    "    if fiveprime\n",
    "        termini[footprint.start] = simulated_fiveprime_termini[i]\n",
    "    elseif threeprime\n",
    "        termini[footprint.stop] = simulated_threeprime_termini[i]\n",
    "    end\n",
    "    hcat(conservation, srna, termini)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = 10000\n",
    "if ~isdir(\"fiveprime_footprints\")\n",
    "    mkdir(\"fiveprime_footprints\")\n",
    "end\n",
    "for n in 1:N\n",
    "    footprint = make_footprint_in_window(n, true, false)\n",
    "    serialize(\"fiveprime_footprints/$(string(n)).bin\", footprint)\n",
    "end\n",
    "\n",
    "if ~isdir(\"threeprime_footprints\")\n",
    "    mkdir(\"threeprime_footprints\")\n",
    "end\n",
    "for n in 1:N\n",
    "    footprint = make_footprint_in_window(n, false, true)\n",
    "    serialize(\"threeprime_footprints/$(string(n)).bin\", footprint)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#background\n",
    "\n",
    "function removefeatures!(fwd_data, rev_data, features)\n",
    "    for f in eachrow(features)\n",
    "        if f.strand == \"+\"\n",
    "            fwd_data[f.start:f.stop] .= -1\n",
    "        else\n",
    "            rev_data[rc(f.stop):rc(f.start)] .= -1\n",
    "        end\n",
    "    end\n",
    "    vcat(fwd_data[fwd_data .>= 0], rev_data[rev_data .>= 0])\n",
    "end\n",
    "\n",
    "function shuffle(data::CircularVector{T}, N::Int) where T <: Real\n",
    "    breakpoints = unique(data)\n",
    "    links = Dict{T, Vector{Int}}()\n",
    "    for bp in breakpoints\n",
    "        links[bp] = findall(==(bp), data) .+ 1\n",
    "    end\n",
    "    samples = Vector{Vector{T}}(undef, N)\n",
    "    for sample in 1:N\n",
    "        idx = rand(1:length(data))\n",
    "        window = Vector{T}(undef, 50)\n",
    "        for i in eachindex(window)\n",
    "            d = data[idx]\n",
    "            window[i] = d\n",
    "            idx = rand(links[d])\n",
    "        end\n",
    "        samples[sample] = window\n",
    "    end\n",
    "    samples\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "foreground = vcat(structuralRNA, footprints_gff)\n",
    "bg_conservation = removefeatures!(fwd_conservation, rev_conservation, foreground)\n",
    "bg_srna = removefeatures!(fwd_srna, rev_srna, foreground)\n",
    "bg_fiveprime = removefeatures!(fwd_fiveprime, rev_fiveprime, foreground)\n",
    "bg_threeprime = removefeatures!(fwd_threeprime, rev_threeprime, foreground)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = 90000\n",
    "background_conservation_samples = shuffle(bg_conservation, N)\n",
    "background_srna_samples = shuffle(bg_srna, N)\n",
    "background_fiveprime_samples = shuffle(bg_fiveprime, N)\n",
    "background_threeprime_samples = shuffle(bg_threeprime, N)\n",
    "\n",
    "if ~isdir(\"fiveprime_backgrounds\")\n",
    "    mkdir(\"fiveprime_backgrounds\")\n",
    "end\n",
    "for n in 1:N\n",
    "    footprint = hcat(background_conservation_samples[n], background_srna_samples[n], background_fiveprime_samples[n])\n",
    "    serialize(\"fiveprime_backgrounds/$(string(n)).bin\", footprint)\n",
    "end\n",
    "\n",
    "if ~isdir(\"threeprime_backgrounds\")\n",
    "    mkdir(\"threeprime_backgrounds\")\n",
    "end\n",
    "for n in 1:N\n",
    "    footprint = hcat(background_conservation_samples[n], background_srna_samples[n], background_threeprime_samples[n])\n",
    "    serialize(\"threeprime_backgrounds/$(string(n)).bin\", footprint)\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.11.5",
   "language": "julia",
   "name": "julia-1.11"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
